evm/p2p/protocol.py

```diff
@@ -0,0 +1,92 @@
+import logging
+import struct
+
+import rlp
+from rlp import sedes
+
+from evm.constants import NULL_BYTE
+
+
+class Command:
+    id = None
+    decode_strict = True
+    structure = []
+
+    @classmethod
+    def encode_payload(cls, data):
+        if isinstance(data, dict):  # convert dict to ordered list
+            if not isinstance(cls.structure, list):
+                raise ValueError("Command.structure must be a list when data is a dict")
+            data = [data[name] for name, _ in cls.structure]
```

Comment:
I would still really like to see validation that we have an exact key match before we reach into the dictionary.  Motivation here is to provide the most helpful error messages possible in the case of a mismatched key.  

Here are two functions copy/pasted from `ethereum-tester` that can be combined to do this.

```python
def validate_no_extra_keys(value, allowed_keys):
    extra_keys = tuple(sorted(set(value.keys()).difference(allowed_keys)))
    if extra_keys:
        raise ValidationError(
            "Only the keys '{0}' are allowed.  Got extra keys: '{1}'".format(
                "/".join(tuple(sorted(allowed_keys))),
                "/".join(extra_keys),
            )
        )


def validate_has_required_keys(value, required_keys):
    missing_keys = tuple(sorted(set(required_keys).difference(value.keys())))
    if missing_keys:
        raise ValidationError(
            "Blocks must contain all of the keys '{0}'.  Missing the keys: '{1}'".format(
                "/".join(tuple(sorted(required_keys))),
                "/".join(missing_keys),
            )
        )
```

Reply:
I'm a bit uncomfortable using those validation functions here, mainly because they raise a ValidationError, which could end up being handled somewhere higher up the stack when in fact a mismatch in the keys here would mean a bug somewhere and we'd want the exception to bubble all the way up. Would you be ok with just comparing both sets of keys and raising a ValueError containing both sets in case they're not identical?